### PR TITLE
Switch to go modules; fix breaking changes

### PR DESCRIPTION
**Pull request by aengelas**
Some of the packages we were using have been pulled, so I had to fix a
call to the github API. TBD whether it breaks things. I _think_ what's
going to happen is that the --force flag is going to stop working, but
I'd need to dig into the api to know for sure.

### CODE REVIEW

**Comment by isobit** [comment]
Indentation is spaces, should be tabs for consistency.

File: deploy.go [hunk 1]
```diff
@@ -257,22 +258,21 @@ func newDeploymentRequest(c *cli.Context, ref string, env string) (*github.Deplo
...
+  if err != nil {
+    return nil, err
+  }
 
 	return &github.DeploymentRequest{
 		Ref:              github.String(ref),
 		Task:             github.String("deploy"),
 		AutoMerge:        github.Bool(false),
 		Environment:      github.String(env),
-		RequiredContexts: contexts,
```
Comment (isobit) [inline deploy.go]: Not passing `RequiredContexts` definitely breaks the force flag. I think `Payload` is just metadata passed to whatever is handling deployments, so we can ignore that. By default (if `RequiredContexts` is nil), GitHub will require all contexts/checks to be successful. It seems to be the force flag disables that by passing an empty list rather than nil, which tells the GitHub API that no contexts should be required.

File: deploy.go [hunk 2]
```diff
@@ -257,22 +258,21 @@ func newDeploymentRequest(c *cli.Context, ref string, env string) (*github.Deplo
...
-		Payload: map[string]interface{}{
-			"force": c.Bool("force"),
-		},
-		// TODO Description:
+		Payload:          github.String(string(payload)),
```
Comment (isobit) [inline deploy.go]: When did this break, and is it documented anywhere? The latest version (v35) godoc has `interface{}` as the `Payload` type, annoyingly. https://pkg.go.dev/github.com/google/go-github/v35@v35.3.0/github#DeploymentRequest

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+go 1.16
+
+require (
+	bitbucket.org/kardianos/osext v0.0.0-20181027061946-15c52d0993e9 // indirect
+	github.com/BurntSushi/toml v0.1.1-0.20141028230038-3883ac1ce943 // indirect
+	github.com/bmizerany/assert v0.0.0-20120716205630-e17e99893cb6 // indirect
+	github.com/codegangsta/cli v1.6.0
+	github.com/fhs/go-netrc v1.0.0 // indirect
+	github.com/github/hub v2.11.2+incompatible
+	github.com/google/go-github v0.0.0-20150126182318-7ea4ee6d2226
```
Comment (isobit) [inline go.mod]: This is hella outdated, thoughts on upgrading in general? Also curious why this introduced breaking chances if it's supposed to be a direct migration to go modules without upgrades.

File: deploy.go (hunk 1 above)
Comment (aengelas) [reply deploy.go]: Yeah, the problem is that in go, an empty slice will be dropped from the JSON when omitempty is set, which it is for DeploymentRequest. This is an api change in the github sdk that (i think) just breaks the way this code is trying to use force.

File: deploy.go (hunk 2 above)
Comment (aengelas) [reply deploy.go]: Oh interesting.

File: go.mod (hunk 3 above)
Comment (aengelas) [reply go.mod]: I had to upgrade some things in order to get the build to work. Some of the transitive dependency packages no longer exist at their original locations.

File: deploy.go (hunk 1 above)
Comment (isobit) [reply deploy.go]: Oh, WTF. So it's possible this never worked correctly? Looks like the vendored source also has `omitempty`.

Comment (isobit) [reply deploy.go]: Apparently `omitempty` does still include the key if the value is a _pointer_ to an empty array: https://play.golang.org/p/pwuCgZaSuUu

Comment (aengelas) [reply deploy.go]: Yes, sorry if that wasn't clear. That's why it used to work, but with github changing their api like this, it will no longer work, afaict. I'm honestly not sure how to disable checks with this new api, but I haven't dug into it. I really just want to get this all working on go modules and fix the auth issues so it's usable again. I think it's a good fallback to have working if slashdeploy fails.

Comment (isobit) [reply deploy.go]: So this is fixed in https://github.com/google/go-github/commit/b2f8d2ec4070d3535b493308749e1c1f5c6c8e55, which we already had in the old vendor: https://github.com/remind101/deploy/blob/4f9b60bed8351ef9bc9aeefa0e2435d19e9e0a55/vendor/github.com/google/go-github/github/repos_deployments.go#L33

So how come this change is using `github.com/google/go-github v0.0.0-20150126182318-7ea4ee6d2226` in `go.mod`? That version is _older_ than what we had in vendor, I think.

Comment (isobit) [reply deploy.go]: It looks like https://github.com/remind101/deploy/blob/4f9b60bed8351ef9bc9aeefa0e2435d19e9e0a55/vendor/vendor.json#L94-L97 is wrong, because https://github.com/google/go-github/blob/7ea4ee6d222607c11ea86e99a6f6723beeae785d/github/repos_deployments.go#L33 doesn't match https://github.com/remind101/deploy/blob/4f9b60bed8351ef9bc9aeefa0e2435d19e9e0a55/vendor/github.com/google/go-github/github/repos_deployments.go#L33

Comment (isobit) [reply deploy.go]: I think it's worth trying to update go-github at least, before this is merged.

File: deploy.go (hunk 2 above)
Comment (aengelas) [reply deploy.go]: Well that was extremely painful, but I've gotten it upgraded.

File: deploy.go [hunk 4]
```diff
@@ -11,10 +12,10 @@ import (
 	"strings"
 	"time"
 
-	"github.com/codegangsta/cli"
 	"github.com/github/hub/git"
 	hub "github.com/github/hub/github"
-	"github.com/google/go-github/github"
+	"github.com/google/go-github/v35/github"
```
Comment (aengelas) [inline deploy.go:17]: This is the most stupid import system I think I've ever used.

Comment (isobit) [reply deploy.go:17]: Actually though wtf, is there not a way to specify this in like the `go.mod`? It has to be in the import??

**Review by isobit — APPROVED** [review]
Looks good, thanks for working though the go-github module craziness. Not a blocker, but can we just get rid of `vendor/` now, or is there a reason we're keeping it around?

File: deploy.go [hunk 5]
```diff
@@ -287,11 +289,11 @@ func isFailed(state string) bool {
 
 // waitState waits for a deployment status that matches the given states, then
 // sends on the returned channel.
-func waitState(states []string, owner, repo string, deploymentID int, c *github.Client) *github.DeploymentStatus {
+func waitState(states []string, owner, repo string, deploymentID int64, c *github.Client) *github.DeploymentStatus {
```
Comment (arlandism) [inline deploy.go:292]: Why the change here to `int64`? I'm assuming this gets deployed to a 64-bit machine so would be 64 bits anyway. I ask because the general best practice is to use the generic int type unless there's a specific reason to pivot.

File: deploy.go [hunk 6]
```diff
@@ -210,7 +212,7 @@ func displayNewCommits(owner string, repo string, ref string, env string, client
 		Environment: env,
 	}
 
-	deployments, _, err := client.Repositories.ListDeployments(owner, repo, opt)
+	deployments, _, err := client.Repositories.ListDeployments(context.TODO(), owner, repo, opt)
```
Comment (arlandism) [inline deploy.go:215]: TIL

File: deploy.go (hunk 5 above)
Comment (isobit) [reply deploy.go:292]: @arlandism I think that's just because the signature of `ListDeploymentStatueses` (used on 296) changed to `int64`

Comment (arlandism) [reply deploy.go:292]: Ah missed that

**Comment by aengelas** [comment]
> can we just get rid of `vendor/` now, or is there a reason we're keeping it around?

We want to vendor our deps a) for stability and b) so we can build in docker without injecting github creds.

File: deploy.go (hunk 4 above)
Comment (aengelas) [reply deploy.go:17]: afaict. as far as go's concerned, this is actually a different package :shrug: go modules seem like one of the worst designed package management systems i've ever seen.
